### PR TITLE
Pepc boost relay implementation

### PR DESCRIPTION
**Pull request by bharath-123**
## 📝 Summary

This PR contains changes related to PEPC-Boost for the MEV relayer.

## ⛱ Motivation and Context

We introduce a new API `/relay/v1/builder/tob_txs`. This API accepts a list of txs for a given slot and parent hash from searchers. It requires searchers to add a payout tx at the end, which transfers eth of value to the relayer. This dictates the value of the txs sent by the TOB searcher. The relayer only stores the highest value TOB txs for a given slot and parent hash. 

We add a new API `/relay/v1/builder/rob_blocks` to which builders sent ROB bids. Upon receiving and successfully validating the sanity of the bid, we pick the highest value TOB txs stored in the relayer for the particular slot and parent hash. We send these TOB txs along with the ROB bid to the payload assembler which assembles these txs and returns the final block to the relayer. 

We ideally want to merge `/relay/v1/builder/rob_blocks` API with `/relay/v1/builder/blocks` since `/relay/v1/builder/blocks` is a special case of a ROB bid without additional TOB txs. But we will keep them separate for development purposes.

W.I.P:

1. We still need to figure out a model for validator and builder payouts.
2. We need to work on more complex state interference checks using tx tracing
## 📚 References

Design Doc: https://docs.google.com/document/d/1DLse86yGLId2idQUIJbaiGxY_RxMxz3LhtdinbA3q4Q/edit

### CODE REVIEW

File: common/types.go [hunk 1]
```diff
@@ -448,6 +450,70 @@ func (e *VersionedExecutionPayload) NumTx() int {
 	return 0
 }
 
+func encodeTransactions(txs []*types.Transaction) [][]byte {
```
Comment (michaelneuder) [inline common/types.go:453]: this file should probably be its own PR with a full set of unit tests. types are a lot of boiler plate but should be really testable

File: database/database.go [hunk 2]
```diff
@@ -25,13 +25,28 @@ type IDatabaseService interface {
 	GetValidatorRegistration(pubkey string) (*ValidatorRegistrationEntry, error)
 	GetValidatorRegistrationsForPubkeys(pubkeys []string) ([]*ValidatorRegistrationEntry, error)
 
+	// Done
```
Comment (michaelneuder) [inline database/database.go:28]: remove these

File: database/database.go [hunk 3]
```diff
@@ -103,6 +118,10 @@ func (s *DatabaseService) prepareNamedQueries() (err error) {
 	(:received_at, :eligible_at, :execution_payload_id, :was_simulated, :sim_success, :sim_error, :sim_req_error, :signature, :slot, :parent_hash, :block_hash, :builder_pubkey, :proposer_pubkey, :proposer_fee_recipient, :gas_used, :gas_limit, :num_tx, :value, :epoch, :block_number, :decode_duration, :prechecks_duration, :simulation_duration, :redis_update_duration, :total_duration, :optimistic_submission)
 	RETURNING id`
 	s.nstmtInsertBlockBuilderSubmission, err = s.DB.PrepareNamed(query)
+	if err != nil {
```
Comment (michaelneuder) [inline database/database.go:121]: this is equivalent to just `return err`

File: database/database_test.go [hunk 4]
```diff
@@ -32,7 +31,8 @@ const (
 )
 
 var (
-	runDBTests   = os.Getenv("RUN_DB_TESTS") == "1" //|| true
+	//runDBTests   = os.Getenv("RUN_DB_TESTS") == "1" //|| true
+	runDBTests   = true
```
Comment (michaelneuder) [inline database/database_test.go:35]: id leave this change only in the local file and not the PR version

File: datastore/redis.go [hunk 5]
```diff
@@ -22,7 +23,7 @@ import (
 var (
 	redisPrefix = "boost-relay"
 
-	expiryBidCache = 45 * time.Second
+	expiryBidCache = 1000 * time.Second
```
Comment (michaelneuder) [inline datastore/redis.go:26]: this is a massive increase. can you justify it? expiry matters a lot b/c it could dramatically change the cost of the redis instance.

File: datastore/redis.go [hunk 6]
```diff
@@ -83,16 +84,27 @@ type RedisCache struct {
...
+	prefixTopTobTxValue string
+	prefixTobTobTx      string
+
 	// prefixes (keys generated with a function)
-	prefixGetHeaderResponse           string
-	prefixExecPayloadCapella          string
-	prefixBidTrace                    string
-	prefixBlockBuilderLatestBids      string // latest bid for a given slot
+	prefixGetHeaderResponse string
+
```
Comment (michaelneuder) [inline datastore/redis.go:93]: can you remove the newlines between each. some clumping would look better IMO

File: datastore/redis_test.go [hunk 7]
```diff
@@ -464,6 +466,88 @@ func _CheckAndSetLastSlotAndHashDeliveredForTesting(r *RedisCache, waitC chan bo
...
+	tx1byte, err := tx1.MarshalBinary()
+	require.NoError(t, err)
+	tx2byte, err := tx2.MarshalBinary()
+	require.NoError(t, err)
+
+	txs := types.Transactions{
+		Transactions: [][]byte{tx1byte, tx2byte},
+	}
+
+	_, err = cache.client.TxPipelined(context.Background(), func(tx redis.Pipeliner) error {
```
Comment (michaelneuder) [inline datastore/redis_test.go:511]: could be nice to check the returns of this ? make sure the values match

File: go.sum
Comment (michaelneuder) [inline go.sum:1]: can you do these in a different PR

File: services/api/service.go [hunk 8]
```diff
@@ -267,20 +297,23 @@ func NewRelayAPI(opts RelayAPIOpts) (api *RelayAPI, err error) {
...
-		beaconClient: opts.BeaconClient,
-		redis:        opts.Redis,
-		memcached:    opts.Memcached,
-		db:           opts.DB,
+		opts:      opts,
+		log:       opts.Log,
+		blsSk:     opts.SecretKey,
+		publicKey: &publicKey,
+		// TODO - make this a config option
+		relayerPayoutAddress: common2.HexToAddress("0x4E9A3d9D1cd2A2b2371b8b3F489aE72259886f1A"),
```
Comment (michaelneuder) [inline services/api/service.go:305]: at least move this to a const at the top of the file

File: services/api/service.go [hunk 9]
```diff
@@ -865,6 +932,7 @@ func (api *RelayAPI) handleRegisterValidator(w http.ResponseWriter, req *http.Re
 		"headSlot":      api.headSlot.Load(),
 		"contentLength": req.ContentLength,
 	})
+	log.Infof("DEBUGGG: handleRegisterValidator request received!!")
```
Comment (michaelneuder) [inline services/api/service.go:935]: remove or rephrase

File: services/api/service.go [hunk 10]
```diff
@@ -1595,6 +1663,620 @@ func (api *RelayAPI) checkBuilderEntry(w http.ResponseWriter, log *logrus.Entry,
...
+		log.WithError(err).Warn("could not set Tob Tx Value")
+		api.RespondError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	api.Respond(w, http.StatusOK, "Tob Tx submitted successfully!")
+	return
+}
+
+func (api *RelayAPI) handleSubmitNewRobBlock(w http.ResponseWriter, req *http.Request) {
```
Comment (michaelneuder) [inline services/api/service.go:1875]: any chance we can use handleSubmitNewBlock out of the box. hard to review when i can't see the diff between this function and the original

File: services/api/service.go [hunk 11]
```diff
@@ -1764,7 +2446,8 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 	// Create the redis pipeline tx
 	tx := api.redis.NewTxPipeline()
 
-	// Reject new submissions once the payload for this slot was delivered - TODO: store in memory as well
+	// bchain - the check for whether the slot of the TOB payload has already been submitted is universal
+	//Reject new submissions once the payload for this slot was delivered - TODO: store in memory as well
```
Comment (michaelneuder) [inline services/api/service.go:2450]: add space before reject

File: services/api/service.go [hunk 12]
```diff
@@ -2061,6 +2744,12 @@ func (api *RelayAPI) handleInternalBuilderStatus(w http.ResponseWriter, req *htt
 	}
 }
 
+func (api *RelayAPI) handleGetHeadSlot(w http.ResponseWriter, req *http.Request) {
+	headSlot := api.headSlot.Load()
+
```
Comment (michaelneuder) [inline services/api/service.go:2749]: newline seems unnecessary

File: services/api/service_test.go [hunk 13]
```diff
@@ -410,6 +415,415 @@ func TestBuilderSubmitBlockSSZ(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestSubmitTobTxs(t *testing.T) {
```
Comment (michaelneuder) [inline services/api/service_test.go]: too long of a test. can you do a TDT here ? looks like a lot of repeated code: https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

**Review by michaelneuder — COMMENTED** [review]
overall first pass. great start. a few comments. similar comments to the builder

- remove commented out stuff and debug statements for reviewability
- can you split this into smaller PRs? where each PR does a specific thing or defines a specific type. if you chain them together it will be so much easier to follow the threads.

File: datastore/redis.go (hunk 5 above)
Comment (bharath-123) [reply datastore/redis.go:26]: nah nah this was for testing.
